services/rpz.c

```diff
@@ -938,107 +1293,692 @@ log_rpz_apply(uint8_t* dname, enum rpz_action a, struct query_info* qinfo,
+	edns->udp_size = EDNS_ADVERTISED_SIZE;
+	edns->ext_rcode = 0;
+	edns->bits &= EDNS_DO;
+	//!inplace_cb_reply_local_call(env, qinfo, NULL, &rep, rcode, edns,repinfo, temp) ||
+	if(!reply_info_answer_encode(qinfo, &rep,
+		*(uint16_t*)sldns_buffer_begin(buf), sldns_buffer_read_u16_at(buf, 2),
+		buf, 0, 0, temp, udpsize, edns, (int)(edns->bits&EDNS_DO), 0)) {
+		error_encode(buf, (LDNS_RCODE_SERVFAIL|BIT_AA), qinfo,
+			*(uint16_t*)sldns_buffer_begin(buf),
+			sldns_buffer_read_u16_at(buf, 2), edns);
+	}
+
+	return 1;
+}
+
+static struct local_rrset*
+rpz_find_synthesized_rrset(int qtype, struct clientip_synthesized_rr* data) {
+	struct local_rrset* cursor = data->data;
+	while( cursor != NULL) {
+		struct packed_rrset_key* packed_rrset = &cursor->rrset->rk;
+		if(htons(qtype) == packed_rrset->type) {
+			return cursor;
+		}
+		cursor = cursor->next;
+	}
+	return NULL;
+}
+
+static void
+rpz_apply_clientip_localdata_action(struct clientip_synthesized_rr* raddr,
+	struct query_info* qinfo, struct edns_data* edns, sldns_buffer* buf,
+	struct regional* temp)
+{
+	struct local_rrset* rrset;
+	enum rpz_action action = RPZ_INVALID_ACTION;
+	struct ub_packed_rrset_key* rp = NULL;
+	int rcode = LDNS_RCODE_NOERROR|BIT_AA;
+	int rrset_count = 1;
+
+	verbose(VERB_ALGO, "rpz: apply client ip trigger: found=%d action=%s",
+		raddr != NULL, rpz_action_to_string(action));
+
+	/* prepare synthesized answer for client */
+
+	action = raddr->action;
+	if(action == RPZ_LOCAL_DATA_ACTION && raddr->data == NULL ) {
+		verbose(VERB_ALGO, "rpz: bug: local-data action but no local data");
+		return;
+	}
+
+	/* check query type / rr type */
+
+	rrset = rpz_find_synthesized_rrset(qinfo->qtype, raddr);
+	if(rrset == NULL) {
+		verbose(VERB_ALGO, "rpz: unable to find local-data for query");
+		rrset_count = 0;
+		goto nodata;
+	}
+
+	rp = respip_copy_rrset(rrset->rrset, temp);
+	if(!rp) {
+		verbose(VERB_ALGO, "rpz: local data action: out of memory");
+		return;
+	}
+
+	rp->rk.flags |= PACKED_RRSET_FIXEDTTL;
+	rp->rk.dname = qinfo->qname;
+	rp->rk.dname_len = qinfo->qname_len;
+nodata:
+	rpz_local_encode(qinfo, edns, buf, temp, rp, rrset_count, rcode);
+}
+
+static inline struct dns_msg*
+rpz_dns_msg_new(struct regional* region)
+{
+	struct dns_msg* msg =
+			(struct dns_msg*)regional_alloc(region,
+							sizeof(struct dns_msg));
+	if(msg == NULL) { return NULL; }
+	memset(msg, 0, sizeof(struct dns_msg));
+
+	return msg;
+}
+
+static inline struct dns_msg*
+rpz_synthesize_nodata(struct rpz* ATTR_UNUSED(r), struct module_qstate* ms)
+{
+	struct dns_msg* msg = rpz_dns_msg_new(ms->region);
+	if(msg == NULL) { return msg; }
+	msg->qinfo = ms->qinfo;
+	msg->rep = construct_reply_info_base(ms->region,
+					     LDNS_RCODE_NOERROR | BIT_RD | BIT_QR | BIT_AA | BIT_RA,
+					     1, //qd
+					     0, //ttl
+					     0, //prettl
+					     0, //expttl
+					     0, //an
+					     0, //ns
+					     0, //ar
+					     0, //total
+					     sec_status_secure);
```

Comment:
Use sec_status_insecure, and also at 1535 and 1563, just like respip.c has sec_status_insecure in 664 (make_new_reply_info).  The message is not one with dnssec, so we can use this sec status.

Reply:
thank you for your review. another question would be if the `reply->is_authoritative` flag should be set as well?

Reply:
Actually that would a great idea.  That flag did not exist in earlier versions where people wanted to 'insert an item in cache', but it does make a lot of sense and make for better output.  I also left review comments on the unbound chat channel that you should look at w.r.t to some connection questions where code goes.

Reply:
This is fixed in commit https://github.com/NLnetLabs/unbound/commit/9f40e93fe6f1473cd2bec93416b397599ce03fc2 .